src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,266 @@
+  MMM_STAT_END = MMM_MAKE_STAT (MMM_MODULE_END)
+} MMM_STATS;
+
+int memmon_add_stat (THREAD_ENTRY *p_thread, MMM_STATS stat, uint64_t size);
+int memmon_sub_stat (THREAD_ENTRY *p_thread, MMM_STATS stat, uint64_t size);
+int memmon_move_stat (THREAD_ENTRY *p_thread, MMM_STATS before_stat, MMM_STATS after_stat, uint64_t size);
+int memmon_resize_stat (THREAD_ENTRY *p_thread, MMM_STATS stat, uint64_t before_size, uint64_t after_size);
+
+namespace cubperf
+{
+  class MMM;
+  class MMM_printer;
+
+  struct MMM_comp_info
+  {
+    int idx;
+    string comp_name;
+    string subcomp_name;
+  }
+
+  typedef struct mmm_mem_stat
+  {
+    std::atomic<uint64_t> init_stat;
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMM_MEM_STAT;
+
+  MMM_comp_info heap_comp_info[] =
+  {
+    {HEAP, "", ""}
+  };
+
+  class MMM_subcomponent
+  {
+    public:
+      MMM_subcomponent (const char *name)
+      {
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* variable */
+      char *m_subcomp_name;
+      atomic<uint64_t> m_cur_stat;
+
+      ~MMM_subcomponent()
+      {
+	delete[] m_subcomp_name;
+      }
+  };
+
+  class MMM_component
+  {
+    public:
+      MMM_component (const char *name)
+      {
+	m_comp_name = new char[strlen (name) + 1];
+	strcpy (m_comp_name, name);
+      }
+
+      /* variable */
+      char *m_comp_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<MMM_subcomponent> m_subcomponent;
+
+      /* function */
+      int add_subcomponent (const char *name);
+
+      ~MMM_component()
+      {
+	delete[] m_comp_name;
+      }
+  };
+
+  class MMM_module
+  {
+    public:
+      MMM_module (const char *name, MMM_comp_info *info);
+      MMM_module() {} /* for dummy */
+
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int m_max_idx = 0x0000FFFF;
+
+      /* variable */
+      char *m_module_name;
+      MMM_MEM_STAT m_stat;
+      vector<MMM_component> m_component;
+      vector<int> m_comp_idx_map;
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+      int add_component (char *name);
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      ~MMM_module()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+  };
+
+  // you have to inheritance "MMM_module" class to your module
+  class MMM_heap_module : public MMM_module
+  {
+    public:
+      heap_stats (const char *name) : template_stats (name) {}
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+
+      ~heap_stats()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+  };
```

Comment:
This is not a scope of this story. Please make PR following the description of story.

Reply:
I deleted it. Thanks.